tools/pythonpkg/src/pyconnection.cpp

```diff
@@ -162,13 +162,14 @@ DuckDBPyConnection *DuckDBPyConnection::RegisterDF(const string &name, py::objec
 	return this;
 }
 
-DuckDBPyConnection *DuckDBPyConnection::RegisterArrow(const string &name, py::object table,
+DuckDBPyConnection *DuckDBPyConnection::RegisterArrow(const string &name, py::object &table,
                                                       const idx_t rows_per_tuple) {
 	if (!connection) {
 		throw std::runtime_error("connection closed");
 	}
-	if (table.is_none() || string(py::str(table.get_type().attr("__name__"))) != "Table") {
-		throw std::runtime_error("Only arrow tables supported");
+	auto py_object_type = string(py::str(table.get_type().attr("__name__")));
+	if (table.is_none() || (py_object_type != "Table" && py_object_type != "FileSystemDataset")) {
+		throw std::runtime_error("Only arrow tables/datasets are supported");
```

Comment:
Why this restriction? In R land we just try to call the export batch stream method and if works we can scan it

Reply:
Well, I could change it to check if the to_batch attr exists.
What I think is nice about this limitation is that we can give a clear error message if something goes wrong.
For example, in the dataset, we can get the to_batch function and call it, but they do not own a row_number attribute. If we did not have this check, this would basically segfault, which I think is pretty bad.

Reply:
why do you need to_batch again?

Reply:
That's what gives us a list of record batches for both tables and datasets.

I checked R-land yesterday, and I think the row_number was just not set from there, so that is why reading from datasets worked there.